source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,693 @@
+// distribute, sublicense, and/or sell copies of the Materials, and to
+// permit persons to whom the Materials are furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Materials.
+//
+// MODIFICATIONS TO THIS FILE MAY MEAN IT NO LONGER ACCURATELY REFLECTS
+// KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+// SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+//    https://www.khronos.org/registry/
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+#include "passes.h"
+
+#include <initializer_list>
+#include <memory>
+#include <tuple>
+#include <unordered_map>
+
+#include "constant_value.h"
+#include "make_unique.h"
+
+namespace spvtools {
+namespace opt {
+
+namespace {
+// The core operate function to get spec operation result with the given opcode
+// and operands in 32 bit words. This function returns the result in 32 bit
+// word.
+uint32_t OperateWords(SpvOp opcode,
+                      const std::vector<uint32_t>& operand_words) {
+  switch (opcode) {
+    // Arthimetics
+    case SpvOp::SpvOpSNegate:
+      // TODO(qining): refer to grammar for such asserts.
+      assert(operand_words.size() == 1u);
+      return -static_cast<int32_t>(operand_words.front());
+    case SpvOp::SpvOpNot:
+      assert(operand_words.size() == 1u);
+      return ~operand_words.front();
+    case SpvOp::SpvOpIAdd:
+      assert(operand_words.size() == 2u);
+      return operand_words.front() + operand_words.back();
+    case SpvOp::SpvOpISub:
+      assert(operand_words.size() == 2u);
+      return operand_words.front() - operand_words.back();
+    case SpvOp::SpvOpIMul:
+      assert(operand_words.size() == 2u);
+      return operand_words.front() * operand_words.back();
+    case SpvOp::SpvOpUDiv:
+      assert(operand_words.size() == 2u);
+      assert(operand_words.back() != 0);
+      return operand_words.front() / operand_words.back();
+    case SpvOp::SpvOpSDiv:
+      assert(operand_words.size() == 2u);
+      assert(operand_words.back() != 0u);
+      return (static_cast<int32_t>(operand_words.front())) /
+             (static_cast<int32_t>(operand_words.back()));
+    case SpvOp::SpvOpSRem: {
+      assert(operand_words.size() == 2u);
+      assert(operand_words.back() != 0u);
+      int32_t a = static_cast<int32_t>(operand_words.front());
+      int32_t b = static_cast<int32_t>(operand_words.back());
+      return a % b;
```

Comment:
Suggest a comment reminding the reader that the sign of a non-zero result comes from operand 1